[docs] Add docs for externalized configuration (#2680)
vividus-framework/vividus: docs/modules/ROOT/pages/tests-configuration.adoc

|===

> == Properties externalization

Reviewer: ```suggestion
== External test configuration points
```

Author: maybe "Test configuration externalization" ? "points" looks a bit weird

Reviewer: "externalization" is too complicated

> 

Reviewer: need to add some general description on when and why it's needed

> . Declare https://en.wikipedia.org/wiki/Environment_variable[environment variable] containing desired property value and then 

Reviewer: ```suggestion
. Declare an https://en.wikipedia.org/wiki/Environment_variable[environment variable] containing the desired property value and then 
```

> point the property to the declared environment variable:

Reviewer: the order of actions here and in examples below is different, but it should be fully aligned

+
[source,properties]
----
bdd.variables.global.input-text=${MY_INPUT_TEXT}
----
+
[source,shell]
----

> export MY_INPUT_TEXT=kekw

Reviewer: we need to be more mindful in our examples

Author: :(

./gradlew runStories
----

> . Pass the property with its value into a test run using command line arguments prepending it with the `-Pvividus.` prefix. 

Reviewer: we need to indicate that this approach is valid for execution via Gradle only

Author: isn't it reflected in the example below?

Reviewer: it's not clear, because users may think it could be used for the embedded sh-script as well, but it's Gradle-only

The prefix only plays role during test start-up phase and cut off during the actual test run:
+
[source,shell]
----
./gradlew runStories -Pvividus.bdd.variables.global.input-text=value1 -Pvividus.bdd.variables.global.output-text=value2
----

> . Generate the `overriding.properties` file on-the-fly and put it into corresponding folder in a test project.

Reviewer: I think we need some example here as well

Author: should I add for both windows and linux platforms?

Reviewer: linux should be enough

|===

== Externalized Configuration

Most common use cases for externalized configuration are tests parametrization with properties passed from CI/CD pipeline, 

> passing of secrets into test run etc.

Reviewer: ```suggestion
passing of secrets into test run etc. The available approaches are listed below.
```

. Declare an https://en.wikipedia.org/wiki/Environment_variable[environment variable] containing the desired property value and 
then point the property to the declared environment variable:
+
[source,shell]
----
export EXTERNAL_CONFIG_VALUE=value
----
+
[source,properties]
----
bdd.variables.global.some-key=${EXTERNAL_CONFIG_VALUE}
----
+

> This approach can be used when tests are started using `./gradlew runStories` command or from `jar` file with tests.

Reviewer: ```suggestion
```
if no limitations, then no reason to add something

. Pass the property with its value into a test run using command line arguments prepending it with the `-Pvividus.` prefix. 
The prefix only plays role during test start-up phase and cut off during the actual test run:
+
[source,shell]
----
./gradlew runStories -Pvividus.bdd.variables.global.some-key-1=value1 -Pvividus.bdd.variables.global.some-key-2=value2
----
+

> This approach can be used only when tests are started using `./gradlew runStories` command.

Reviewer: ```suggestion
This approach can be used only when tests are run via Gradle.
```

. Generate the `overriding.properties` file on-the-fly and put it into corresponding folder in a test project.
+

> The following script assumes that we are at the root of the test project folder.

Reviewer: ```suggestion
The following script assumes that we are in the root of the test project folder.
```

+
[source,shell]
----
cat > src/main/resources/overriding.properties << EOD
bdd.variables.global.some-key-1=value1
bdd.variables.global.some-key-2=value2
EOD
----
+

> This approach can be used when tests are started using `./gradlew runStories` command.

Reviewer: ```suggestion
This approach can be used when tests are run via Gradle.
```

http.auth.password=ENC(WVnXhzG704KRZZFYL7hfpA4YoSm8V2F5eNXyr7CRYCichteHvNS3nB23zaI8CuLy)
----

> === How to externalize an encryptor password in VIVIDUS

Reviewer: too specific, this approach is relevant for any property

http.auth.password=ENC(WVnXhzG704KRZZFYL7hfpA4YoSm8V2F5eNXyr7CRYCichteHvNS3nB23zaI8CuLy)
----

> Please see <<_externalized_configuration>> for encryptor secret externalization.

Reviewer: ```suggestion
Please see <<_externalized_configuration>> to find the ways how encryptor password can be passed to the tests.
```
